### PR TITLE
Implement Operator Selection as a separate activity

### PR DESCRIPTION
**Pull request by gorandostanic**

### CODE REVIEW

File: connect/src/com/telenor/connect/ConnectSdk.java [hunk 1]
```diff
@@ -142,17 +131,26 @@ private static Intent getAuthIntent(Map<String, String> parameters) {
...
+        sdkProfile.onStartAuthorization(
+                parameters,
+                getUiLocales(),
+                new SdkProfile.OnStartAuthorizationCallback() {
             @Override
-            public void onSuccess() {
-                Intent intent = getAuthIntent(parameters);
-                intent.putExtra(ConnectUtils.CUSTOM_LOADING_SCREEN_EXTRA, customLoadingLayout);
-                activity.startActivityForResult(intent, requestCode);
+            public void onSuccess(Intent nextIntent) {
```
Comment (atanasdb) [inline connect/src/com/telenor/connect/ConnectSdk.java]: wrong indent

**Review by atanasdb — COMMENTED** [review]
To be honest that's a bit too hard for me to follow.
 
I can understand that by abstracting the Activity and Fragment classes you achieve code reuse, but does it necessarily has to be in this patch? That makes the PR really huge and complex (at least for my level). Maybe you can introduce the activity in this patch, and then do the refactoring in another one? Also if you introduce a regression, it'd be a bit hard to say was it because of the refactoring or because of the newly introduced functionality.

Also I'm not convinced that this is the right way to set up callbacks. E.g. in handleOperatorDiscoveryFailure when you callback.onSuccess(intent); to go back to ConnectSdk::authenticate makes the code hard to follow. Why not just pass this callback to the OperatorSelectionActivity and call if/when it finishes successfully. That makes it very easy to reason when ConnectSdk.authenticate callback is called  - when everything is present, and we're ready to go to the /authorize endpoint. And you will also be able to remove one of the onStartAuthorization methods in MobileConnectSdkProfile.

File: connect/src/com/telenor/mobileconnect/ui/OperatorSelectionActivity.java [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+                mccAndMnc[0],
+                mccAndMnc[1],
+                uri.getQueryParameter(SUBSCRIBER_ID_QUERY_PARAM_NAME),
+                new SdkProfile.OnStartAuthorizationCallback() {
+                    @Override
+                    public void onSuccess(Intent nextIntent) {
+                        if (!isConnectActivity(nextIntent)) {
+                            throw new RuntimeException("Wrong intent.");
+                        }
+                        Activity activity = OperatorSelectionActivity.this;
```
Comment (atanasdb) [inline connect/src/com/telenor/mobileconnect/ui/OperatorSelectionActivity.java]: What?
Why don't just startActivityForResult?
Why do you need the this magic

File: connect/src/com/telenor/mobileconnect/ui/OperatorSelectionActivity.java [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+
+
+
+    @Override
+    public void onError(Object errorData) {
+        cancelAndShowErrorMessage(this);
+    }
+
+    @Override
+    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
```
Comment (atanasdb) [inline connect/src/com/telenor/mobileconnect/ui/OperatorSelectionActivity.java]: This is very wrong order for me.
So basically what happens here is OperatorSelection OnResult is called with the results of the ConnectActivity activity - the tokens.
The correct way should be that - ConnectActivity calls startActivityForResult(OperatorSelectionActivity) then waits for the results, and proceed.

File: connect/src/com/telenor/connect/ConnectSdk.java (hunk 1 above)
Comment (gorandostanic) [reply connect/src/com/telenor/connect/ConnectSdk.java]: Done!

File: connect/src/com/telenor/mobileconnect/ui/OperatorSelectionActivity.java (hunk 2 above)
Comment (gorandostanic) [reply connect/src/com/telenor/mobileconnect/ui/OperatorSelectionActivity.java]: Done!

File: connect/src/com/telenor/mobileconnect/ui/OperatorSelectionActivity.java (hunk 3 above)
Comment (gorandostanic) [reply connect/src/com/telenor/mobileconnect/ui/OperatorSelectionActivity.java]: The suggested order as such is even more wrong, because it would require the Connect Activity to know about Operator Discovery Activity, whilst I think the other way around is more natural. IMO Connect Activity should be MC agnostic, since it's used by both profiles unchanged.

To tell the complete truth, I could subclass ConnectActivity with something named MobileConnectActivity doing exactly that... is that what you want me to do?

Comment (atanasdb) [reply connect/src/com/telenor/mobileconnect/ui/OperatorSelectionActivity.java]: No, I just don't think that the OperatorSelection activity has to receive tokens. In Android the logic typically is, if you need to fetch some info (let's say contact) you will launch ContactActivity, (the user selects it) you set the result there, call finish() and handle the result in the original activity - the one that asked for the contact.

In your case, the way u've implemented it looks like OperatorSelection asks ConnectActivity for tokens, and receives them. Which is wrong, right? The way I see things is since ConnectActivity needs an authorization_url to proceed, it either have it (in case of ConnectId) or it need to fetch it from somewhere (OperatorDiscovery). In the second case, it'll call OperatorSelection activity for help.

In pseudo code, something like:

ConnectSdk::authenticate always calls 
app's signinactivity.startforresult(connect_activity)
   
 connect_activity: onStart -> 1) if authorization_url is known (that is if mcc/mnc are present or if it is ConnectId) -> proceed as usual;
                               else 2) connect_activity.startforresult(os_selection)

in case of 2) , os_selection does it job, call finish(). and then ,

 connect_activity: onActivityResult (called from os_selection' finish) -> grab the authorization_url and proceed as in the previous case

Comment (atanasdb) [reply connect/src/com/telenor/mobileconnect/ui/OperatorSelectionActivity.java]: "Connect Activity to know about Operator Discovery Activity" is not necessarily true.
e.g. in ConnectSDK in case of an unknown authorization_url, you could have:
ConnectSdk.getSdkProfile().getActivityForFindingAuthorizationUrl(), which for ConnectId might just throw some RuntimeException since you are not supposed to call that, and for MobileConnect it can give you the OperatorSelection activity. But in any case, if you fix the "wiring" of the callbacks, and avoid the "Abstract" mentioned above in this PR, that should be a really one nice, small and sweet PR :) (in theory ;) )

Comment (gorandostanic) [reply connect/src/com/telenor/mobileconnect/ui/OperatorSelectionActivity.java]: What's happening here is basically an example of the delegation pattern, since OperatorSelectionActivity started in authenticate() promises the same result as ConnectActivity. So, if I merely rename OperatorSelectionActivity to something like ConnectActivityWithOperatorSelectionDelegate, would your doubts be somewhat relieved?  

The truth is that we do not need OperatorSelection as a entirely stand-alone activity (who would ever use it without performing authentication?). That OperatorSelection thing is, whichever way you look at it, a part of authorisation flow... but only sometimes (i.e. when we are not able to fetch endpoints without user interaction). In other cases, it is not even that :)

Comment (gorandostanic) [reply connect/src/com/telenor/mobileconnect/ui/OperatorSelectionActivity.java]: Or, maybe better: ConnectActivityDelegateWithOperatorSelection?

Comment (atanasdb) [reply connect/src/com/telenor/mobileconnect/ui/OperatorSelectionActivity.java]: No, I don't think they'll be. Because I think you have unnecessary dependencies, which complicate the code. In your case you have:

**ConnectSDK.authenticate()** can invoke **connect_activity**  
**ConnectSDK.authenticate()** can invoke **delegate**
**delegate** invokes **connect_activity**
**connect_activity** returns results to **delegate**
**delegate** returns results to **main_activity**
**connect_activity** returns results to **main_activity**


compare with:
**ConnectSDK.authenticate()** can invoke **connect_activity** 
**connect_activity** can invoke **operator_selection_activity** 
 **operator_selection_activity**  returns result to **connect_activity**
**connect_activity** returns results to **main_activity**

Comment (atanasdb) [reply connect/src/com/telenor/mobileconnect/ui/OperatorSelectionActivity.java]: And it's not only that you have more dependencies, but there's code to make this communication between dependencies possible, which I think is not necessary.

Comment (atanasdb) [reply connect/src/com/telenor/mobileconnect/ui/OperatorSelectionActivity.java]: Ok, as I've done something quick for the Friday hack, just to show what I tried to say here:
https://github.com/telenordigital/connect-android-sdk/pull/57

A few things:

1. connect_activity is started right away. then it's responsibility to fetch:
    - in case of Connect Id - well-known config
    - in case of Mobile Connect - get mcc/mnc and then well-known, or if mcc/mnc fails, launch operator selection to get mcc/mnc and then well-known. I believe the code is 
easier to follow and it's less

2. Prefer using string literals for things that used only once

3. I've moved error handling to the client apps in the SigninActivity. I don't think the SDK needs to mess with the UI for the errors visualisation and that must be up to the app developer to decide.
   
4. I use anonymous classes as much as possible and put the callbacks definition as close as possible to where they are needed.

Comment (gorandostanic) [reply connect/src/com/telenor/mobileconnect/ui/OperatorSelectionActivity.java]: I have skimmed through your PR (https://github.com/telenordigital/connect-android-sdk/pull/57 ):

I think it's OK, at least for not being an "official" PR, but the main intention of my PR was to keep the old Connect Id code as much unchanged as possible (and with as little "glue" needed to reconcile Connect Id and MC as possible), in order to enable the Connect Id team to maintain their stuff without worrying too much about MC. In other words, the plan of my PR was to implement MC merely as a "client" of Connect Id, since both profiles do exactly the same thing after a certain point in the flow has been reached. This plan is kind of broken in your PR, since ConnectActivity and ConnectSdkProfile (and possibly some other classes) implement stuff solely motivated by MC.

I think the design patterns used in both PRs are equally legitimate. This PR implements a variant of delegate pattern, while yours implements something I cannot put a naming label on, but it's obvious that it's OK. The main difference is that my PR tries to keep the differences between Connect Id and MC in the com.telenordigital.mobileconnect package, while yours doesn't impose such restrictions upon itself.

I would have had some more pointed comments if this PR had been "official". For example, I didn't like the sudden "publicness" of ConnectSdk.getAuthorizeUriAndSetLastAuthState(). Similarly, I didn't like setValue() and getValue in SdkProfle interface neither, since these two methods are obviously motivated by some implementation needs, not by the need for establishing a valid contract. Nevertheless, since your PR is still a rough and a non-official one, I will refrain from reviewing it in details.

Comment (atanasdb) [reply connect/src/com/telenor/mobileconnect/ui/OperatorSelectionActivity.java]: Yeah, it's idea was to illustrate a new way of connecting OperatorSelection and ConnectActivity. Should it become "official" as you say, it needs a bit more refactoring for the things you mentioned. I see nothing wrong with getAuthorizeUriAndSetLastAuthState being public, if it was only for the getAuthorizeUri part, so maybe have two separate methos. For setValue/getValue I agree...

Comment (gorandostanic) [reply connect/src/com/telenor/mobileconnect/ui/OperatorSelectionActivity.java]: I think getAuthorizeUriAndSetLastAuthState() is too nitty-gritty to be public. For example, even if we disregard the side effect like setting sLastAuthState for a moment, there is an obvious bug there right now: if a developer calls it before the profile has been properly initialised (which is legitimate, since the method is public), he will get a null pointer exception. This means that one must know when to call this method, because not all the instances of the time are created equal => method is not suitable for being public.
